Fix file resolution issues

circlingthesun opened the pull request:
I've fixed some of the issues I've been experiencing.

1. The ```join_path``` function prefixed all paths with a slash. This is problematic as it resulted it some relative paths resolving the the file system root and raising an error.

2. The ```get_relative_path``` path function assumed that ```base``` is at the root of ```path```. When truncating ```base.length``` characters off ```path``` an empty string would result if the path was shorter than the base.

3. I've changed the behaviour of the ```noStartingSlash``` parameter in ```get_relative_path``` so that it results in the path having no starting slash. The previous behaviour made no sense to me. Correct me if I misunderstood something.

4. I've also made some style fixes that my linter picked up. Triple equals and brackets around if statements mostly.

I've not managed to get things 100% working for my use case, but this irons out some of the more obvious issues.

circlingthesun commented:
I've resolved some more issues in this update:

1. Javascript references were not being resolved when they did not have extensions.
2. Some arbitrary substitutions of JS references were being made and was not picked up by the test. I fixed the bug and the test. JS references now need to be surrounded by quotes.
3. Short JS references had file extensions added to them which breaks SystemJS. This no longer happens. A test was added.

circlingthesun commented:
There are still some unresolved issues I haven't gotten to yet. When replacing references to javascript files without extensions is easy to get things wrong. Consider ```var home = require('home')```. ```lib``` is pretty generic so its easy to substitute incorrect references. I've mitigated the issue a bit by restricting javascript file substitutions to strings inside quotes. A better solution would be to find ```require```,  ```define```, ```System.register```, etc calls and only substitute extensionless references in those places.

smysnk commented:
I guess my question is "home" a realistic reference that gulp-rev-all should be trying to match as a file.  One of the things I was trying avoid in the re-write is having some funky long regex to try and match require/ define / etc statements.  I think any string that contains `./`, `../` or `directory/abc` can be safely matched to a file, but I would argue single word includes like `"home"` are a little bit dangerous and could definitely lead to sale false-positives.  For the example of "home", what would the actual path of this file be?  Is it not a script file that is basically in the root directory called "/home.js"? I don't think `gulp-rev-all` should be trying to tell you where to put your scripts but at same time what is the lengths we'll go to allow this to work safely?

circlingthesun commented:
Well, "home.js" was just one example I encountered. What makes this issue more common than it would otherwise be, is libraries such as Angular that does dependency injection. If I create a module called 'xyz', the file is likely to end being called 'xyz.js'. I will then end with this kind of code:

```js
require('xyz');

angular.controller('myController', ['xyz', function(xyz){
   ...
}]);
```

Which will result in:

```js
require('xyz.123');

angular.controller('myController', ['xyz.123', function(xyz){
   ...
}]);
```

As you see the 2nd xyz substitution is incorrect. It doesn't really matter what string 'xyz' is, people like to often use the same string for the filename and module name.

It is a lot of overhead to do regex parsing for each type of module system. It might be a good idea to write some type of hook that people can use to control substitutions. A user could provide a function that takes a file content as a parameter and returns the annotated file contents that indicates what parts are safe to substitute.

Something like:

```js

var annotate = function(content) {
  ...
  ...
   return [
     {content: '...', jsSubstituteSafe: true},
     {content: '...', jsSubstituteSafe: false},
  ]
}

revAll({annotateContent: annotate});

```

Revall can then just substitute js file names the on the strings that are marked safe, and then join the parts together. This could be extended to work on other types of problematic substitutions too.

Thoughts?

circlingthesun commented:
I've implemented my suggestion and added some docs. This effectively offloads the issue to the end user in cases when rev-all does the wrong substitution.

circlingthesun commented:
I've managed to fix all the issues I experienced. I recently started switching from RequireJS to SystemJS so I've tested this with both module loaders and two module formats (CommonJS & AMD).  I've had to use regular expressions to identify which parts of my code base is safe to do shorthand javascript substitutions on. Having slashes in my modules to prevent bad substitutions was not quite working out. I used the new hooks to achieve this without adding magical regular expressions to rev-all. Without the external code substitutions are 99% correct as JS module references now need to be quoted. I've included the code below.

```js
var annotator = function(contents, path){

    var moduleLoaderRegex = /(?:define|require|System\.register|System\.import)\s*\(\s*(?:(?:['"][^'"]*['"]\s?,\s?)?(?:\[[^\]]*|(?:function))|(?:['"][^'"]*['"]\s?))/g;
    var safeShortSubList = [];
    var result;
    
    while(result = moduleLoaderRegex.exec(contents)){
        if(!result){
            continue;
        }
        safeShortSubList.push(result[0]);
    }

    var fragments = [{'contents': contents}];

    for(var i = 0; i < fragments.length; i++){

        for(var j = 0; j < safeShortSubList.length; j++){
            
            var parts = fragments[i].contents.split(safeShortSubList[j]);
            
            if(parts.length === 1){
                continue;
            }

            var args = [i, 1];

            for(var idx = 0; idx < parts.length; idx++){
                if(parts[idx]){
                    args.push({'contents': parts[idx]});
                }

                var isLast = (idx === parts.length-1);
                if(!isLast){
                    args.push({'contents': safeShortSubList[j], 'safeShort': true});
                }
            }

            // Nothing changed
            if(args.length < 4){
                continue;
            }

            Array.prototype.splice.apply(fragments, args);
        }
        
    }

    return fragments;
};

var replacer = function(fragment, replaceRegExp, newReference, referencedFile){

    var regExp;

    if(!referencedFile.path.match(/.js$/) || fragment.safeShort){
        regExp = replaceRegExp;
    } else {
        // Make the extension manditory
        var regExpParts = replaceRegExp.toString().split('/');
        var opts = regExpParts.pop();
        regExpParts.splice(0, 1); // Remove the leading slash
        var regexStr = regExpParts.join('\/').replace('(\\.js)?', '(\\.js)');
        regExp = new RegExp(regexStr, opts);
    }

    fragment.contents = fragment.contents.replace(regExp, '$1' + newReference + '$3$4');
    
};

```
